src/lib/components/forms/DocumentUpload.svelte

```diff
@@ -0,0 +1,458 @@
+  let projects: Project[] = [];
+
+  let loading = false;
+  let uploader: HTMLInputElement;
+
+  let fileDropActive: boolean;
+
+  const ocrEngineOptions = [
+    {
+      value: "tess4",
+      label: "Tesseract",
+      help: $_("uploadOptions.tesseract"),
+    },
+    {
+      value: "textract",
+      label: "Textract",
+      help: $_("uploadOptions.textract"),
+    },
+  ];
+
+  let ocrEngine = ocrEngineOptions[0];
+
+  $: csrf_token = $page.data.csrf_token;
+  $: projects = $page.data.projects.results;
+
+  $: total = files.reduce((t, file) => {
+    return t + file.size;
+  }, 0);
+
+  function addFiles(filesToAdd: FileList) {
+    files = files.concat(Array.from(filesToAdd).filter(isSupported));
+  }
+
+  function removeFile(index: number) {
+    files = files.filter((f, i) => i !== index);
+  }
+
+  function formatFileType(filetype: string) {
+    if (filetype && filetype.includes("/")) {
+      return filetype.split("/")[1];
+    }
+    return "unknown";
+  }
+
+  function filenameToTitle(filename: string): string {
+    const [name, ...ext] = filename.split(".");
+    return name.replace(/_/g, " ");
+  }
+
+  // handle uploads client side instead of going through the server
+  async function onSubmit(e: SubmitEvent) {
+    loading = true;
+    const form = e.target as HTMLFormElement;
+    const fd = new FormData(form);
+
+    const result = await upload(fd, csrf_token);
+
+    // send data up
+    await applyAction(result);
+
+    if (result.type === "success") {
+      form.reset();
+      files = [];
+    }
+
+    loading = false;
+  }
+
+  afterUpdate(() => {
+    const dt = new DataTransfer();
+
+    files.forEach((file) => {
+      dt.items.add(file);
+    });
+
+    uploader.files = dt.files;
+  });
+</script>
+
+<form
+  method="post"
+  enctype="multipart/form-data"
+  action="/app/upload/"
+  on:submit|preventDefault={onSubmit}
+>
+  <Flex gap={1} align="stretch" wrap>
+    <div class="files">
+      <!-- add any header and messaging using this slot -->
+      <slot />
+
+      <div class="fileList" class:empty={files.length === 0}>
+        {#each files as file, index}
+          <Flex align="center" gap={1}>
+            <p class="fileInfo">
+              {formatFileType(file.type)} / {filesize(file.size)}
+            </p>
+            <div class="title">
+              <Text name="title" value={filenameToTitle(file.name)} required />
+              <input type="hidden" name="filename" value={file.name} />
+            </div>
+            <button
+              class="fileRemove"
+              on:click|preventDefault={() => removeFile(index)}
+            >
+              <XCircleFill24 />
+            </button>
+          </Flex>
+        {:else}
+          <Empty icon={File24}>
+            Get started by selecting, pasting or dragging-and-dropping files
+            below
+          </Empty>
+        {/each}
+
+        <div class="total">
+          <p>Total upload size: {filesize(total)}</p>
+        </div>
```

Comment:
I'll take a pass on styling and placement for the total size counter. If there's a limit that we can exceed, we should message that and disable the submit button when it's over.

Reply:
The limits are a little complicated:
- 500 mb for PDFs
- 25 mb for other file types
- No total upload limit, except what will fit in your browser memory

We're doing everything in the browser now, so really the only limit is on individual files.

Reply:
Cool, since we have the type and the size in the list item above, makes sense to add a warning or error message if the size is too big for the type.